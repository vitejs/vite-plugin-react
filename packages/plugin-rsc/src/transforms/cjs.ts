import type { Program, Node } from 'estree'
import MagicString from 'magic-string'
import { analyze } from 'periscopic'
import { walk } from 'estree-walker'

export function transformCjsToEsm(
  code: string,
  ast: Program,
): { output: MagicString } {
  const output = new MagicString(code)
  const analyzed = analyze(ast)

  const parentNodes: Node[] = []
  const hoistedCodes: string[] = []
  let hoistIndex = 0
  walk(ast, {
    enter(node) {
      parentNodes.push(node)
      if (
        node.type === 'CallExpression' &&
        node.callee.type === 'Identifier' &&
        node.callee.name === 'require' &&
        node.arguments.length === 1
      ) {
        let isTopLevel = true
        for (const parent of parentNodes) {
          if (
            parent.type === 'FunctionExpression' ||
            parent.type === 'FunctionDeclaration' ||
            parent.type === 'ArrowFunctionExpression'
          ) {
            isTopLevel = false
          }
          // skip locally declared `require`
          const scope = analyzed.map.get(parent)
          if (scope && scope.declarations.has('require')) {
            return
          }
        }

        if (isTopLevel) {
          // top-level scope `require` to dynamic import
          // (this allows handling react development/production re-export within top-level if branch)
          output.update(node.start, node.callee.end, '(await import')
          output.appendRight(node.end, ')')
        } else {
          // hoist non top-level `require` to top-level
          const hoisted = `__cjs_to_esm_hoist_${hoistIndex}`
          const importee = code.slice(
            node.arguments[0]!.start,
            node.arguments[0]!.end,
          )
          hoistedCodes.push(`const ${hoisted} = await import(${importee});\n`)
          output.update(node.start, node.end, hoisted)
          hoistIndex++
        }
      }
    },
    leave() {
      parentNodes.pop()!
    },
  })
  for (const hoisted of hoistedCodes.reverse()) {
    output.prepend(hoisted)
  }
  output.prepend(`const exports = {}; const module = { exports };\n`)
  return { output }
}
