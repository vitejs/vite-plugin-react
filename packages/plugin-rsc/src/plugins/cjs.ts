import { parseAstAsync, type Plugin } from 'vite'
import { parseIdQuery } from './utils'
import { findClosestPkgJsonPath } from 'vitefu'
import path from 'node:path'
import fs from 'node:fs'
import * as esModuleLexer from 'es-module-lexer'
import { transformCjsToEsm } from '../transforms/cjs'

export function cjsModuleRunnerPlugin(): Plugin[] {
  // use-sync-external-store is known to work fine so don't show warning
  const warnedPackages = new Set<string>(['use-sync-external-store'])

  return [
    {
      name: 'cjs-module-runner-transform',
      apply: 'serve',
      applyToEnvironment: (env) => env.config.dev.moduleRunnerTransform,
      async transform(code, id) {
        if (
          id.includes('/node_modules/') &&
          !id.startsWith(this.environment.config.cacheDir) &&
          /\b(require|exports)\b/.test(code)
        ) {
          id = parseIdQuery(id).filename
          if (!/\.[cm]?js$/.test(id)) return

          // skip genuine esm
          if (id.endsWith('.mjs')) return
          if (id.endsWith('.js')) {
            const pkgJsonPath = await findClosestPkgJsonPath(path.dirname(id))
            if (pkgJsonPath) {
              const pkgJson = JSON.parse(
                fs.readFileSync(pkgJsonPath, 'utf-8'),
              ) as { type?: string }
              if (pkgJson.type === 'module') return
            }
          }

          // skip faux esm (e.g. from "module" field)
          const [, , , hasModuleSyntax] = esModuleLexer.parse(code)
          if (hasModuleSyntax) return

          // warning once per package
          const packageKey = extractPackageKey(id)
          if (!warnedPackages.has(packageKey)) {
            this.warn(
              `Found non-optimized CJS dependency in '${this.environment.name}' environment. ` +
                `It is recommended to add the dependency to 'environments.${this.environment.name}.optimizeDeps.include'.`,
            )
            warnedPackages.add(packageKey)
          }

          const ast = await parseAstAsync(code)
          const result = transformCjsToEsm(code, ast)
          const output = result.output
          // TODO: can we use cjs-module-lexer to properly define named exports?
          // for re-exports, we need to eagerly transform dependencies though.
          // https://github.com/nodejs/node/blob/f3adc11e37b8bfaaa026ea85c1cf22e3a0e29ae9/lib/internal/modules/esm/translators.js#L382-L409
          output.append(`
;__vite_ssr_exportAll__(module.exports);
export default module.exports;
`)
          return {
            code: output.toString(),
            map: output.generateMap({ hires: 'boundary' }),
          }
        }
      },
    },
  ]
}

function extractPackageKey(id: string): string {
  // .../.yarn/cache/abc/... => abc
  const yarnMatch = id.match(/\/.yarn\/cache\/([^/]+)/)
  if (yarnMatch) {
    return yarnMatch[1]!
  }
  // .../node_modules/@x/y/... => @x/y
  // .../node_modules/x/... => x
  if (id.includes('/node_modules')) {
    id = id.split('/node_modules/').at(-1)!
    let [x, y] = id.split('/')
    if (x!.startsWith('@')) {
      return `${x}/${y}`
    }
    return x!
  }
  return id
}
