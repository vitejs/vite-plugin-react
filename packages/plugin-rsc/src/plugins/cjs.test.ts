import { test, expect, describe } from 'vitest'
import { cjsModuleRunnerPlugin, cjsModuleRunnerTransform } from './cjs'
import { createServer } from 'vite'

describe.skip(cjsModuleRunnerTransform, () => {
  test('basic', async () => {
    const input = `
exports.ok = true;
`
    const output = await cjsModuleRunnerTransform(input)
    expect(output).toMatchSnapshot()
    const server = await createServer({
      configFile: false,
      plugins: [
        {
          name: 'test-entry',
          resolveId(source) {
            if (source === 'virtual:test') {
              return '\0' + source
            }
          },
          load(id) {
            if (id === '\0virtual:test') {
              return output
            }
          },
        },
      ],
    })
    const mod = await server.ssrLoadModule('virtual:test')
    expect(mod).toMatchInlineSnapshot(`
      {
        "ok": true,
      }
    `)
  })

  test('basic', async () => {
    const input = `
if (process.env.NODE_ENV === 'production') {
  module.exports = require('./cjs/use-sync-external-store.production.js');
} else {
  module.exports = require('./cjs/use-sync-external-store.development.js');
}
`
    const output = await cjsModuleRunnerTransform(input, {
      define: {
        'process.env.NODE_ENV': JSON.stringify('development'),
      },
    })
    expect(output).toMatchSnapshot()
    const server = await createServer({
      configFile: false,
      plugins: [cjsModuleRunnerPlugin()],
    })
    const mod = await server.ssrLoadModule(
      'use-sync-external-store/shim/index.js',
    )
    expect(mod).toMatchInlineSnapshot(`
      {
        "useSyncExternalStore": [Function],
      }
    `)
  })
})
