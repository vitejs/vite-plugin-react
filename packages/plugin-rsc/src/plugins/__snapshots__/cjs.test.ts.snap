// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cjsModuleRunnerTransform > basic 1`] = `
"//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function() {
	return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
//#region \\0virtual:entry-inner
var require___virtual_entry_inner = /* @__PURE__ */ __commonJS({ "\\\\0virtual:entry-inner": ((exports) => {
	exports.ok = true;
}) });

//#endregion
//#region \\0virtual:entry
var import___virtual_entry_inner = /* @__PURE__ */ __toESM(require___virtual_entry_inner());
__vite_ssr_exportAll__(import___virtual_entry_inner.default);

//#endregion"
`;

exports[`cjsModuleRunnerTransform > basic 3`] = `
"import m from "./cjs/use-sync-external-store.development.js";

//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function() {
	return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
//#region \\0virtual:require-to-import/./cjs/use-sync-external-store.development.js
var require_use_sync_external_store_development = /* @__PURE__ */ __commonJS({ "\\\\0virtual:require-to-import/./cjs/use-sync-external-store.development.js": ((exports, module) => {
	module.exports = m;
}) });

//#endregion
//#region \\0virtual:entry-inner
var require___virtual_entry_inner = /* @__PURE__ */ __commonJS({ "\\\\0virtual:entry-inner": ((exports, module) => {
	module.exports = require_use_sync_external_store_development();
}) });

//#endregion
//#region \\0virtual:entry
var import___virtual_entry_inner = /* @__PURE__ */ __toESM(require___virtual_entry_inner());
__vite_ssr_exportAll__(import___virtual_entry_inner.default);

//#endregion"
`;
